### PR TITLE
[REEF-1205] Move submission-time parameters into AvroYarnJobSubmissio…

### PR DESCRIPTION
**Pull request by afchung**
…nParameters

This addressed the issue by
  * Removing DriverConfigurationOptions.MaxApplicationSubmissions and moved it into the IJobSubmission object.
  * Move parameters from AppSubmissionParameters to JobSubmissionParameters.
  * Modified tests and examples.

JIRA:
  [REEF-1205](https://issues.apache.org/jira/browse/REEF-1205)

### CODE REVIEW

**Comment by afchung** [comment]
@anupam128 Please have a look when you have time. Thanks!

**Comment by anupam128** [comment]
@afchung Sure will take a look.

File: lang/cs/Org.Apache.REEF.Client/API/IJobSubmission.cs [hunk 1]
```diff
@@ -57,6 +57,11 @@ public interface IJobSubmission
         int DriverMemory { get; }
 
         /// <summary>
+        /// The maximum amount of time an application can be submitted.
```
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmission.cs]: nit: maximum number of times?

**Comment by anupam128** [comment]
Is this a breaking change for users? How is public API impacted in terms of the way users were specifying these values?

**Comment by afchung** [comment]
@anupam128 Overall, only one thing has changed for the user, which is the way they specify the maximum number of submissions. It was erroneously in `DriverConfiguration` when it is really a submission parameter. Furthermore, it is only mainly used in the scenario of Driver Restarts. Provided that it is not really a widely used function of REEF, I propose that we skip the deprecation phase for this, 
I have tested this on a YARN cluster for both submission methods.
Thanks!

**Comment by afchung** [comment]
@anupam128 I've addressed your comments, please have another look. Thanks!

**Comment by anupam128** [comment]
@markusweimer can you please confirm that we can skip the deprecation phase here?
Rest of the code LGTM.

**Comment by markusweimer** [comment]
Was this API ever a part of a release? If not, all changes are fair game.

**Comment by afchung** [comment]
@markusweimer I believe it was part of last release.

**Comment by markusweimer** [comment]
In that case, we might have to do the deprecation route. Is it possible to offer both parameters, with the new one dominating if both are set?

**Comment by afchung** [comment]
@anupam128 @markusweimer I've reverted the behavior of `MaxApplicationSubmissions` to be backward compatible for the old REEF client and deprecated it. Thanks!

File: lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs [hunk 2]
```diff
@@ -71,11 +71,16 @@ public interface IJobSubmissionBuilder
...
         /// <summary>
-        /// Set driver memory in megabytes
+        /// Set driver memory in megabytes.
         /// </summary>
         IJobSubmissionBuilder SetDriverMemory(int driverMemoryInMb);
 
         /// <summary>
+        /// Set the maximum amount of times a job can be submitted.
+        /// </summary>
+        IJobSubmissionBuilder SetMaxApplicationSubmissions(int maxAppSubmissions);
```
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs]: @markusweimer adding this method is a breaking change on this interface. I am not sure about how people use it. Is this fine?

Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs]: I do not believe anyone else implements this interface either.

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs]: Yeah, only we implement it. It's ok to add methods.

**Comment by anupam128** [comment]
I will try to test and merge.

**Pull request closed without merging**